### PR TITLE
Enable upgrade of Kubernetes clusters with master VMs using managed disks

### PR DESCRIPTION
**Pull request by amanohar**
**What this PR does / why we need it**:
Introduction of Kubernetes master VMs to start using managed disks both for OS and etcddisk broke upgrade operation:
1) Upgrade operation was supported on unmanaged disks only and making manageddisk the default StoragrProfile broke upgrade operation OOB.
2) Switch to using managed disk also started using a default disk name to be assigned by Disk RP (vs. one assigned by ACS Engine like for other resources). This could be problematic in many ways:
     - With main one being relying on and understanding DiskRP’s naming convention to discover the right etcd disk for each master VM. This might not be a big issue in the RP because disk names/ids can be saved in the database but from ACS Engine standpoint for operations to be idempotent disk names need to be deterministic. 
     - This  also adds unnecessary complexity of loading and editing the template during upgrade with the disk name generated by Disk RP.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, 1) will close that issue when PR gets merged)*: fixes #
1) This PR adds support to enable upgrade of clusters using managed disk master VMs.
2) The code has been updated to use a deterministic name for etcd disks. However, any cluster created between June 22nd to until this PR gets merged still gets non-deterministic names for etcd disks.
3) Change upgrade template to have a managedDisk section when attaching an existing etcddisk

Pending fixes:
1) Supported attaching of auto generated etcd disks (names) during upgrade.

**Special notes for your reviewer**:

**Release note**:

```release-note
```

---
This change is [Reviewable](https://reviewable.io/reviews/azure/acs-engine/1008)

### CODE REVIEW

File: pkg/armhelpers/interfaces.go [hunk 1]
```diff
@@ -65,6 +66,9 @@ type ACSEngineClient interface {
 	// RBAC
 	CreateRoleAssignment(scope string, roleAssignmentName string, parameters authorization.RoleAssignmentCreateParameters) (authorization.RoleAssignment, error)
 	CreateRoleAssignmentSimple(applicationID, roleID string) error
+
+	// MANAGED DISKS
+	DeleteManagedDisk(resourceGroupName string, diskName string, cancel <-chan struct{}) (<-chan disk.OperationStatusResponse, <-chan error)
```
Comment (colemickens) [inline pkg/armhelpers/interfaces.go]: Whaaaaa? I thought MD disks were auto-deleted with lifetime of VM? :(

**Comment by seanknox** [comment]
Hey @amanohar all feature and bug work needs a [corresponding issue](https://github.com/Azure/acs-engine/blob/master/CONTRIBUTING.md#issue-lifecycle) so we can track and work into an upcoming release.

**Comment by seanknox** [comment]
@acs-bot test this please

File: pkg/armhelpers/interfaces.go (hunk 1 above)
Comment (amanohar) [inline pkg/armhelpers/interfaces.go]: Yes, MD don't get deleted with the VM.

**Comment by amanohar** [comment]
@acs-bot test this please

File: pkg/armhelpers/mockclients.go [hunk 2]
```diff
@@ -265,3 +266,26 @@ func (mc *MockACSEngineClient) CreateRoleAssignment(scope string, roleAssignment
...
+	go func() {
+		defer func() {
+			close(errChan)
+		}()
+		defer func() {
+			close(respChan)
+		}()
+		errChan <- nil
+		respChan <- disk.OperationStatusResponse{}
+		time.Sleep(1 * time.Second)
```
Comment (colemickens) [inline pkg/armhelpers/mockclients.go]: Can we remove this? I think I have a commit in one of the PRs I just merged that removes the sleeps from this file. AFAICT there's no need for them, and removing it just sped up the tests a bit.

**Comment by amanohar** [comment]
@acs-bot test this please

**Comment by seanknox** [comment]
giphy-downsized 3

**Pull request merged**